kafka-client/src/main/java/dev/responsive/kafka/internal/db/rs3/RS3KVFlushManager.java

```diff
@@ -156,64 +161,131 @@ public CompletionStage<RemoteWriteResult<Integer>> flush() {
+    }
+
+    Optional<Long> writeWalSegmentSync(List<WalEntry> entries) {
+      return rs3Client.writeWalSegment(
+          storeId,
+          lssId,
+          pssId,
+          expectedWrittenOffset,
+          endOffset,
+          entries
+      );
+    }
+
+  }
+
+  private static final class RS3KVWriter implements RemoteWriter<Bytes, Integer> {
+    private final RS3StreamFactory streamFactory;
+    private final RS3KVTable table;
+    private final int kafkaPartition;
+    private final List<WalEntry> retryBuffer = new ArrayList<>();
+    private final StreamSenderMessageReceiver<WalEntry, Optional<Long>> sendRecv;
+
+    private RS3KVWriter(
+        final UUID storeId,
+        final RS3Client rs3Client,
+        final RS3KVTable table,
+        final int pssId,
+        final LssId lssId,
+        final long endOffset,
+        final Optional<Long> expectedWrittenOffset,
+        final int kafkaPartition
+    ) {
+      this.table = Objects.requireNonNull(table);
+      this.streamFactory = new RS3StreamFactory(
+          storeId,
+          rs3Client,
+          pssId,
+          lssId,
+          endOffset,
+          expectedWrittenOffset
+      );
+      this.kafkaPartition = kafkaPartition;
+      this.sendRecv = streamFactory.writeWalSegmentAsync();
     }
 
     long endOffset() {
-      return endOffset;
+      return streamFactory.endOffset;
     }
 
     @Override
     public void insert(final Bytes key, final byte[] value, final long timestampMs) {
-      streamSender.sendNext(table.insert(kafkaPartition, key, value, timestampMs));
+      maybeSendNext(table.insert(kafkaPartition, key, value, timestampMs));
     }
 
     @Override
     public void delete(final Bytes key) {
-      streamSender.sendNext(table.delete(kafkaPartition, key));
+      maybeSendNext(table.delete(kafkaPartition, key));
+    }
+
+    private void maybeSendNext(WalEntry entry) {
+      retryBuffer.add(entry);
+      ifActiveStream(sender -> sender.sendNext(entry));
+    }
+
+    private void ifActiveStream(Consumer<StreamSender<WalEntry>> streamConsumer) {
+      if (sendRecv.isActive()) {
+        try {
+          streamConsumer.accept(sendRecv.sender());
+        } catch (final RS3TransientException e) {
+          // Retry the stream in flush()
+        }
+      }
     }
 
     @Override
     public CompletionStage<RemoteWriteResult<Integer>> flush() {
-      streamSender.finish();
-      return resultFuture.thenApply(
-          flushedOffset -> {
-            LOG.debug("last flushed offset for pss/lss {}/{} is {}", pssId, lssId, flushedOffset);
-            return RemoteWriteResult.success(kafkaPartition);
-          }
-      );
+      ifActiveStream(StreamSender::finish);
+
+      return sendRecv.receiver().handle((result, throwable) -> {
+        Optional<Long> flushedOffset = result;
+        if (throwable instanceof RS3TransientException) {
+          flushedOffset = streamFactory.writeWalSegmentSync(retryBuffer);
```

Comment:
the retries for this call happen in the grpc client right?

Reply:
Yes, that's right. I was trying to keep retry logic encapsulated in the grpc client. But there's not really a way I could think of to encapsulate retries for the async API.